test/datetime_test.exs

```diff
@@ -165,15 +173,46 @@ defmodule DatetimeTest do
              ])
   end
 
-  test "datetime2", context do
-    assert nil == Types.encode_datetime2(nil)
-    enc = Types.encode_datetime2(@datetime)
-    assert ~N[2015-04-08 15:16:23.000000] == Types.decode_datetime2(7, enc)
-    enc = Types.encode_datetime2(@datetime_fsec)
-    assert ~N[2015-04-08 15:16:23.123456] == Types.decode_datetime2(7, enc)
-    enc = Types.encode_datetime2({@date, {131, 56, 23, 0}}, 0)
-    assert {@date, {131, 56, 23, 0}} == Types.decode_datetime2(0, enc)
+  describe "decode_datetime2" do
+    test "truncates microseconds with scale of 7 to the max of 6" do
+      encoded = Types.encode_datetime2(@datetime2)
+
+      assert Types.decode_datetime2(7, encoded) ==
+               ~N[2015-04-08 15:16:23.123456]
+    end
+
+    test "with 0 microseconds" do
+      enc = Types.encode_datetime2(@datetime)
+      assert Types.decode_datetime2(7, enc) == ~N[2015-04-08 15:16:23.000000]
+    end
+
+    test "with scale of 6" do
+      enc = Types.encode_datetime2(@datetime_fsec)
+      assert Types.decode_datetime2(7, enc) == ~N[2015-04-08 15:16:23.123456]
+    end
+
+    test "with scale of 0" do
+      enc = Types.encode_datetime2({@date, {13, 56, 23, 0}}, 0)
+      assert Types.decode_datetime2(0, enc) == ~N[2015-04-08 13:56:23]
```

Comment:
I think this is not what TDS should return, instead of decoding datetime to elixir NaiveDateTime or DateTime we should still keep TDS more generic and keep only tuples. Then in ecto 2.x, tds_ecto  or ecto_sql mssql adapter there should be `loader` and `dumper` that should convert from and to tuple to `:naive_datetime`, `:naive_datetime_usec`... so ecto can handle them. 

If we do this, then we can introduce Tds.DateTime, Tds.DateTime2 and Tds.DateTimeOffset as custom ecto types to add support for 7th digit precision for users that need it. This also should be case for migrations, if column is defined in migration as :navie_datetime_usec without precision option it should default to 6. If someone need 7 digits, then it should be explicit setup in migration file.

Do you agree?

Reply:
Yeah, this makes sense to keep the datetime's normalized as tuples and let the Ecto library, Tds.DateTime2, other custom Ecto types handle the final Ecto type encoding/decoding.  I had a custom type for Datetime2, which did the truncating and can help out with that.